aiographite/aiographitesend.py

```diff
@@ -0,0 +1,449 @@
+import asyncio
+from aiographite.graphite_escaping import metrics_name_to_graphite, metrics_name_from_graphite
+import pickle
+import struct
+import socket
+import os
+import time
+
+
+DEFAULT_GRAPHITE_PLAINTEXT_PORT = 2003
+DEFAULT_GRAPHITE_PICKLE_PORT = 2004
+SUPPORT_PROTOCOLS = ["plaintext", "pickle"]
+
+
+class AioGraphiteSendException(Exception):
+    pass
+
+
+class AsyncioGraphiteSendService(object):
+
+	def __init__(self, graphite_server, graphite_port, protocol = "pickle"):
+
+		self.graphite_server_address = (graphite_server, graphite_port)
+
+		self.connect_to_graphite()
+
+		self.protocol = protocol
+
+		self.loop = asyncio.get_event_loop()
+
+
+	def connect_to_graphite(self):
+		"""
+			Connect to Graphite Server based on Provided Server Address
+		"""
+		try:
+			self.socket = socket.create_connection(self.graphite_server_address)
+			self.socket.setblocking(False)
+		except socket.gaierror:
+			raise AioGraphiteSendException("Unable to connect to the provided server address %s:%s" % self.graphite_server_address)
+		except Exception as e:
+			raise e
+		return self.socket
+
+
+
+	def send_single_data(self, metric_dir_list, value, timestamp = None):
+		"""
+			@example: 
+				Assuming that 
+
+					Expected_Metric_Name  =  metaccounts.authentication.password.attempted
+
+				Then input metric_dir_list should be
+
+					metric_dir_list = [metaccounts, authentication, password, attempted]
+
+			@metric_dir_list: List of string
+			@timestamp: the type should be int
+
+			If you're very confident that the metric name is valid, then use <method: send_single_valid_data> instead.
+
+		"""		
+		valid_metric_name = self.to_graphite_valid_metric_name(metric_dir_list)
+		self.send_single_valid_data(valid_metric_name, value, timestamp)
+
+
+
+	def send_dataset_list(self, dataset, timestamp = None):
```

Comment:
I see a lot of combinational methods here. Specifically combinations of:

1. dataset submissions (list vs single)
2. conversions from the various inputs.

If you try to support all permutations of this, it's an N^2 problem. Instead, we should focus on a bunch of smaller, composable methods that we can string together as necessary.

e.g.

```
metric_names = [self.to_graphite_metric_list(metric_parts) for metric_parts in metric_list]
self.send_to_graphite(metric_names)
```

vs 

```
self.send_metric_names_to_graphite(metric_list)
```